File: server/src/main/com/broll/gainea/server/core/battle/BattleResult.java
package com.broll.gainea.server.core.battle;

import com.broll.gainea.server.core.map.Location;
import com.broll.gainea.server.core.objects.Unit;
import com.broll.gainea.server.core.player.Player;
import com.google.common.collect.Lists;

import java.util.ArrayList;
import java.util.List;

public class BattleResult extends BattleContext {

    private boolean retreated;

    private boolean attackersWon;
    private boolean defendersWon;

    public BattleResult(boolean retreated, BattleContext context) {
        super(context.attackers, context.defenders);
        this.retreated = retreated;
        this.attackersWon = !retreated && hasSurvivingAttackers() && !hasSurvivingDefenders();
        this.defendersWon = retreated || (hasSurvivingDefenders() && !hasSurvivingAttackers());
    }

    public List<Player> getWinningPlayers() {
        return getNonNeutralOwners(getWinnerOwners());
    }

    public List<Player> getLosingPlayers() {
        return getNonNeutralOwners(getLoserOwners());
    }

    public boolean isWinner(Player player) {
        return getWinnerOwners().contains(player);
    }

    public boolean isLoser(Player player) {
        return getLoserOwners().contains(player);
    }

    public boolean isNeutralWinner() {
        return isNeutralOwner(getWinnerOwners());
    }

    public boolean isNeutralLoser() {
        return isNeutralOwner(getLoserOwners());
    }

    public Location getAttackerEndLocation() {
        return attackersWon() ? getLocation() : getSourceLocation();
    }

    public boolean attackersWon() {
        return attackersWon;
    }

    public boolean defendersWon() {
        return defendersWon;
    }

    private List<Player> getWinnerOwners() {
        if (attackersWon()) {
            return Lists.newArrayList(attackingPlayer);
        } else if (defendersWon()) {
            return defendingPlayers;
        }
        return new ArrayList<>();
    }

    private List<Player> getLoserOwners() {
        if (attackersWon()) {
            return defendingPlayers;
        } else if (defendersWon()) {
            return Lists.newArrayList(attackingPlayer);
        }
        return new ArrayList<>();
    }

    public boolean attackerRetreated() {
        return retreated;
    }

    public boolean isDraw() {
        return !attackersWon() && !defendersWon();
    }

    public List<Unit> getWinnerUnits() {
        if (attackersWon()) {
            return getAttackers();
        } else if (defendersWon()) {
            return getDefenders();
        }
        return new ArrayList<>();
    }

    public List<Unit> getLoserUnits() {
        if (attackersWon()) {
            return getDefenders();
        } else if (defendersWon()) {
            return getAttackers();
        }
        return new ArrayList<>();
    }

    public List<Unit> getWinnerSurvivingUnits() {
        if (attackersWon()) {
            return getAliveAttackers();
        } else if (defendersWon()) {
            return getAliveDefenders();
        }
        return new ArrayList<>();
    }

    public List<Unit> getLoserSurvivingUnits() {
        if (attackersWon()) {
            return getAliveDefenders();
        } else if (defendersWon()) {
            return getAliveAttackers();
        }
        return new ArrayList<>();
    }

    public List<Unit> getWinnerDeadUnits() {
        if (attackersWon()) {
            return getKilledAttackers();
        } else if (defendersWon()) {
            return getKilledDefenders();
        }
        return new ArrayList<>();
    }

    public List<Unit> getLoserDeadUnits() {
        if (attackersWon()) {
            return getKilledDefenders();
        } else if (defendersWon()) {
            return getKilledAttackers();
        }
        return new ArrayList<>();
    }

    public Location getEndLocation(Player player) {
        if (isAttacker(player)) {
            return getAttackerEndLocation();
        } else if (isDefender(player)) {
            return location;
        }
        return null;
    }

    public List<Player> getKillingPlayers(Unit unit) {
        if (isAttacking(unit)) {
            return getDefendingPlayers();
        } else if (isDefending(unit)) {
            return Lists.newArrayList(getAttackingPlayer());
        }
        return new ArrayList<>();
    }
}

File: server/src/main/com/broll/gainea/server/core/battle/BattleResult.kt
package com.broll.gainea.server.core.battle

import com.broll.gainea.server.core.map.Location
import com.broll.gainea.server.core.objects.IUnit
import com.broll.gainea.server.core.player.Player

class BattleResult(val retreated: Boolean, context: BattleContext) :
    AbstractBattleContext<UnitSnapshot>(
        context.attackers.map { UnitSnapshot(it) },
        context.defenders.map { UnitSnapshot(it) },
        context.rounds
    ) {
    val attackersWon: Boolean = !retreated && hasSurvivingAttackers() && !hasSurvivingDefenders()
    val defendersWon: Boolean = retreated || hasSurvivingDefenders() && !hasSurvivingAttackers()

    val winningPlayers: List<Player>
        get() = getNonNeutralOwners(winnerOwners)
    val losingPlayers: List<Player>
        get() = getNonNeutralOwners(loserOwners)

    fun isWinner(player: Player) = winnerOwners.contains(player)


    fun isLoser(player: Player) = loserOwners.contains(player)

    val isNeutralWinner: Boolean
        get() = isNeutralOwner(winnerOwners)
    val isNeutralLoser: Boolean
        get() = isNeutralOwner(loserOwners)
    val attackerEndLocation: Location
        get() = if (attackersWon) location else sourceLocation

    private val winnerOwners: List<Player>
        get() {
            if (attackersWon) {
                return attackingPlayers
            } else if (defendersWon) {
                return defendingPlayers
            }
            return listOf()
        }
    private val loserOwners: List<Player>
        get() {
            if (attackersWon) {
                return defendingPlayers
            } else if (defendersWon) {
                return attackingPlayers
            }
            return listOf()
        }

    val isDraw: Boolean
        get() = !attackersWon && !defendersWon
    val winnerUnits: List<UnitSnapshot>
        get() {
            if (attackersWon) {
                return attackers
            } else if (defendersWon) {
                return defenders
            }
            return listOf()
        }
    val loserUnits: List<UnitSnapshot>
        get() {
            if (attackersWon) {
                return defenders
            } else if (defendersWon) {
                return attackers
            }
            return listOf()
        }
    val winnerSurvivingUnits: List<UnitSnapshot>
        get() {
            if (attackersWon) {
                return aliveAttackers
            } else if (defendersWon) {
                return aliveDefenders
            }
            return listOf()
        }
    val loserSurvivingUnits: List<UnitSnapshot>
        get() {
            if (attackersWon) {
                return aliveDefenders
            } else if (defendersWon) {
                return aliveAttackers
            }
            return listOf()
        }
    val winnerDeadUnits: List<UnitSnapshot>
        get() {
            if (attackersWon) {
                return killedAttackers
            } else if (defendersWon) {
                return killedDefenders
            }
            return listOf()
        }
    val loserDeadUnits: List<UnitSnapshot>
        get() {
            if (attackersWon) {
                return killedDefenders
            } else if (defendersWon) {
                return killedAttackers
            }
            return listOf()
        }

    fun getDamageDealt(player: Player) =
        rounds.flatMap { it.damageTaken }.count { it.source.owner == player }

    fun getEndLocation(player: Player): Location? {
        if (isAttacker(player)) {
            return attackerEndLocation
        } else if (isDefender(player)) {
            return location
        }
        return null
    }

    fun getKillingPlayers(unit: IUnit): List<Player> {
        if (killedAttackers.any { it.id == unit.id }) {
            return defendingPlayers
        } else if (killedDefenders.any { it.id == unit.id }) {
            return attackingPlayers
        }
        return listOf()
    }
}
